## Capstone Final Project - Weeks 4 and 5 

### 1. Business Problem

In this final project, I am going to find an optimal location for a gym in Toronto. It´s specifically targeted to stakeholders interested in opening a sport center in Toronto, Canada. 

As there are many gyms located in Toronto, the main idea is to detect locations, which have few or none gyms. Moreover, I´ll analyze proximity to Yoga Places and Parks, as they are possible competitors to a gym.

I´ll base my analysis on data science tools to generate a list of neighborhoods determining pros and cons so that the stakeholder can choose the final location.


### 2. Data Sources

Having presented the main problem we want to solve, the main factors which will affect the stakeholders´ decision are:


•	number of gyms in each neighborhood,

•	number of Yoga Places and Parks.


Following data sources will be needed to extract/generate the required information:


•	Toronto´s Neighborhoods data will be obtained from Wikipedia.

•	number of gyms and their location in every neighborhood will be obtained using Foursquare API.

•	number of Yoga Places / Parks and their location in every neighborhood will be obtained using Foursquare API.


### 3. Methodology

#### 3.1 Dependences Needed

Before we get the data and start exploring it, we need to download the dependencies needed.

In [16]:
import requests

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Solving environment: ...working... done

# All requested packages already installed.

Solving environment: ...working... done

# All requested packages already installed.

Libraries imported.


#### 3.2 Downloading and Exploring Data

For the Toronto neighborhood data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. We will scrape the page and wrangle the data, clean it, and then read it into a pandas dataframe to work with it. 

In [17]:
wiki = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wiki_page = requests.get(wiki)
wiki_raw = pd.read_html(wiki_page.content, header=0)[0]
wiki_raw.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

We drop all the Postal Codes which are not assigned to a specific Neighborhood.

In [18]:
df = wiki_raw[wiki_raw.Neighborhood != 'Not assigned']
df.reset_index(inplace = True)
df.drop(['index'],axis=1,inplace=True)
df.head(5)

C:\Users\Diego\Anaconda3\lib\site-packages\pandas\core\frame.py:3694: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [19]:
df.groupby(['Postal Code'])

In [20]:
df.shape

(103, 3)

#### 3.3 Additioning more data

In a Csv file, I have the information regarding the geographical coordinates of each postal code.
I will download and work with the data and merge it to the table we are already creating.

In [21]:
url = 'http://cocl.us/Geospatial_data'
df_geo = pd.read_csv(url)
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [22]:
df_geo.shape

(103, 3)

In [23]:
df = df.join(df_geo.set_index('Postal Code'),on='Postal Code')
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

In [24]:
df.shape

(103, 5)

#### 3.4 Creating a Map

I will create a map of Toronto with neighborhoods superimposed on top.

In [25]:
!conda install -c conda-forge geocoder --yes
import geocoder
from geopy.geocoders import Nominatim 

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

Solving environment: ...working... done

# All requested packages already installed.

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [26]:
# create map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_Toronto)  
    
map_Toronto

#### 3.5 Introducing Foursquare API 

Now, I´m going to start utilizing the Foursquare API to explore the neighborhoods.

In [28]:
CLIENT_ID = 'JJ5P2Q0RAZHW4GWXXFZE40KHAT0NYQBS20IOXAUET1YHLA2R' # your Foursquare ID
CLIENT_SECRET = '0LZOBNHAGOXDBVIDV0TPQ3KTJEOEAHMU4MWR1XCWSZQCKUL2' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: JJ5P2Q0RAZHW4GWXXFZE40KHAT0NYQBS20IOXAUET1YHLA2R
CLIENT_SECRET:0LZOBNHAGOXDBVIDV0TPQ3KTJEOEAHMU4MWR1XCWSZQCKUL2


In [29]:
neighborhood_latitude = df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


In [30]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=JJ5P2Q0RAZHW4GWXXFZE40KHAT0NYQBS20IOXAUET1YHLA2R&client_secret=0LZOBNHAGOXDBVIDV0TPQ3KTJEOEAHMU4MWR1XCWSZQCKUL2&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [31]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f125919ea795307ab6aea9e'},
 'response': {'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.751976046055574,
          'lng': -79.33214044722958}],
        'distance': 245,
        'cc': 'CA'

In [32]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [33]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0             Brookbanks Park               Park  43.751976 -79.332140
1              TTC stop #8380           Bus Stop  43.752672 -79.326351
2               Variety Store  Food & Drink Shop  43.751974 -79.333114
3  TTC stop - 44 Valley Woods           Bus Stop  43.755402 -79.333741

In [34]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [35]:
toronto_venues = getNearbyVenues(names=df['Neighborhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [36]:
print(toronto_venues.shape)
toronto_venues.head()

(2163, 7)


Neighbourhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3         Parkwoods              43.753259              -79.329656   
4  Victoria Village              43.725882              -79.315572   

                        Venue  Venue Latitude  Venue Longitude  \
0             Brookbanks Park       43.751976       -79.332140   
1              TTC stop #8380       43.752672       -79.326351   
2               Variety Store       43.751974       -79.333114   
3  TTC stop - 44 Valley Woods       43.755402       -79.333741   
4      Victoria Village Arena       43.723481       -79.315635   

      Venue Category  
0               Park  
1           Bus Stop  
2  Food & Drink Shop  
3           Bus Stop  
4       Hockey Arena

In [37]:
toronto_venues.groupby('Neighbourhood').count()

Neighborhood Latitude  \
Neighbourhood                                                               
Agincourt                                                               4   
Alderwood, Long Branch                                                  7   
Bathurst Manor, Wilson Heights, Downsview North                        23   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      28   
Berczy Park                                                            57   
Birch Cliff, Cliffside West                                             5   
Brockton, Parkdale Village, Exhibition Place                           22   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  13   
Cedarbrae                                                               8   
Central Bay Street                                                     66   
Christie                                                               16   
Church and Wellesley                                                   72   
Clarks Corners, Tam O'Shanter, Sullivan                                12   
Cliffside, Cliffcrest, Scarborough Village West                         4   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills                                                              28   
Dorset Park, Wexford Heights, Scarborough Town ...                      6   
Downsview                                                              17   
Dufferin, Dovercourt Village                                           16   
East Toronto, Broadview North (Old East York)                           4   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                     10   
Fairview, Henry Farm, Oriole                                           65   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Glencairn                                                               4   
Golden Mile, Clairlea, Oakridge                                         9   
Guildwood, Morningside, West Hill                                       9   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
Hillcrest Village                                                       4   
Humber Summit                                                           2   
Humberlea, Emery                                                        2   
Humewood-Cedarvale                                                      4   
India Bazaar, The Beaches West                                         22   
Kennedy Park, Ionview, East Birchmount Park                             4   
Kensington Market, Chinatown, Grange Park                              66   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                                       14   
Lawrence Park                                                           4   
Leaside                                                                33   
Little Portugal, Trinity                                               44   
Malvern, Rouge                                                          1   
Milliken, Agincourt North, Steeles East, L'Amor.

#### 3.6 Analyzing each Neighborhood

In [38]:

# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2         Parkwoods                  0        0                   0   
3         Parkwoods                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Basketball Court  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  \
0           0                   0          0       0              0   
1           0                   0          0       0              0   
2           0                   0          0       0              0   
3           0                   0          0       0              0   
4           0                   0          0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bridal Shop  Bubble Tea Shop  Building  Burger Joint  Burrito Place  \
0            0                0         0             0              0   
1            0                0         0             0              0   
2            0                0         0             0              0   
3            0                0         0             0              0   
4            0                0         0             0   

In [39]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Lounge  Airport Service  \
0      0.0                 0.0             0.0              0.0   
1      0.0                 0.0             0.0              0.0   
2      0.0                 0.0             0.0              0.0   
3      0.0                 0.0             0.0              0.0   
4      0.0                 0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0               0.0             0.000000           0.0       0.0          0.0   
1               0.0             0.000000           0.0       0.0          0.0   
2               0.0             0.000000           0.0       0.0          0.0   
3               0.0             0.000000           0.0       0.0          0.0   
4               0.0             0.035714           0.0       0.0          0.0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0         0.0                  0.0               0.0                 0.0   
1         0.0                  0.0               0.0                 0.0   
2         0.0                  0.0               0.0                 0.0   
3         0.0                  0.0               0.0                 0.0   
4         0.0                  0.0               0.0                 0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery      Bank  Bar  \
0            0.0        0.0         0.0         0.0     0.0  0.000000  0.0   
1            0.0        0.0         0.0         0.0     0.0  0.000000  0.0   
2            0.0        0.0         0.0         0.0     0.0  0.086957  0.0   
3            0.0        0.0         0.0         0.0     0.0  0.250000  0.0   
4            0.0        0.0         0.0         0.0     0.0  0.000000  0.0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0             0.0               0.0               0.0                 0.0   
1             0.0               0.0               0.0                 0.0   
2             0.0               0.0               0.0                 0.0   
3             0.0               0.0               0.0                 0.0   
4             0.0               0.0               0.0                 0.0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0    0.0              0.0       0.0         0.0                 0.0   
1    0.0              0.0       0.0         0.0                 0.0   
2    0.0              0.0       0.0         0.0                 0.0   
3    0.0              0.0       0.0         0.0                 0.0   
4    0.0              0.0       0.0         0.0                 0.0   

   Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0        0.0     0.0            0.0        0.0  0.000000   
1        0.0     0.0            0.0        0.0  0.000000   
2        0.0     0.0            0.0        0.0  0.000000   
3        0.0     0.0            0.0        0.0  0.000000   
4        0.0     0.0            0.0        0.0  0.035714   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                   0.0            0.25      0.0     0.000000   
1                   0.0            0.00      0.0     0.000000   
2                   0.0            0.00      0.0     0.043478   
3                   0.0            0.00      0.0     0.000000   
4                   0.0            0.00      0.0     0.000000   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0              0.0       0.0           0.0            0.0       0.0  

Now, we are going to analyze the most common venues in each Neighbourhood.


In [41]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [48]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighbourhood   1st Most Common Venue  \
0                                           Agincourt                  Lounge   
1                              Alderwood, Long Branch             Pizza Place   
2     Bathurst Manor, Wilson Heights, Downsview North                    Bank   
3                                     Bayview Village                    Bank   
4                   Bedford Park, Lawrence Manor East             Coffee Shop   
5                                         Berczy Park             Coffee Shop   
6                         Birch Cliff, Cliffside West         College Stadium   
7        Brockton, Parkdale Village, Exhibition Place                    Café   
8   Business reply mail Processing Centre, South C...      Light Rail Station   
9   CN Tower, King and Spadina, Railway Lands, Har...          Airport Lounge   
10                                Caledonia-Fairbanks                    Park   
11              Canada Post Gateway Processing Centre             Coffee Shop   
12                                          Cedarbrae        Hakka Restaurant   
13                                 Central Bay Street             Coffee Shop   
14                                           Christie           Grocery Store   
15                               Church and Wellesley             Coffee Shop   
16            Clarks Corners, Tam O'Shanter, Sullivan             Pizza Place   
17    Cliffside, Cliffcrest, Scarborough Village West            Intersection   
18                     Commerce Court, Victoria Hotel             Coffee Shop   
19                                         Davisville             Pizza Place   
20                                   Davisville North    Gym / Fitness Center   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn    Fast Food Restaurant   
22                                          Don Mills                     Gym   
23  Dorset Park, Wexford Heights, Scarborough Town...       Indian Restaurant   
24                                          Downsview                    Park   
25                       Dufferin, Dovercourt Village                  Bakery   
26      East Toronto, Broadview North (Old East York)                    Park   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...                    Café   
28                       Fairview, Henry Farm, Oriole          Clothing Store   
29             First Canadian Place, Underground city             Coffee Shop   
30    Forest Hill North & West, Forest Hill Road Park                   Trail   
31                           Garden District, Ryerson          Clothing Store   
32                                          Glencairn             Pizza Place   
33                    Golden Mile, Clairlea, Oakridge                  Bakery   
34                  Guildwood, Morningside, West Hill              Restaurant   
35  Harbourfront East, Union Station, Toronto Islands             Coffee Shop   
36                      High Park, The Junction South                    Café   
37                                  Hillcrest Village                 Dog Run   
38                                      Humber Summit                     Gym   
39                                   Humberlea, Emery  Furniture / Home Store   
40                                 Humewood-Cedarvale                   Field   
41                     India Bazaar, The Beaches West                    Park   
42        Kennedy Park, Ionview, East Birchmount Park       Convenience Store   
43          Kensington Market, Chinatown, Grange Park                    Café   
44  Kingsview Village, St. Phillips, Martin Grove ...                    Park   
45                   Lawrence Manor, Lawrence Heights          Clothing Store   
46                                      Lawrence Park                    Park   
47                                            Leaside     Sporting Goods Shop   
48                           Little Portugal, Trinity                  

### 4. Results and Discussion

The main goal of the project was to analyze the city of Toronto and find a variety of neighborhoods which would be suitable to open a gym. The main driver would be to find a place which did not have competition.

With Data Science tools we first understood which Neighborhoods had Gyms or Fitnesses Centers as one of the main top 10 venues. We found 14 (representing 15% of Toronto´s Neighborhoods). 

Having taken out just 15% of the neighborhoods was the first step. We decided to focus afterwards in other venues which could replace a Gym such as:

•	Tennis Court,
•	Skating Rink,
•	Yoga Studio,
•	Swim School,
•	Climbing Gym, among others.

We found that 57 neighborhoods (60% of the total) had these venues within the top 10.
After taking those out, we end up with our final group of possible 14 neighborhood where to open a Gym (25% of the total). In these neighborhoods, there will not be much competition in terms of other gyms nor other venues, which can replace it.

The final list is the following:
•	Bedford Park, Lawrence Manor East
•	Berczy Park
•	CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
•	Central Bay Street
•	Clarks Corners, Tam O'Shanter, Sullivan
•	Dufferin, Dovercourt Village
•	Fairview, Henry Farm, Oriole
•	Garden District, Ryerson
•	Guildwood, Morningside, West Hill
•	High Park, The Junction South
•	Lawrence Manor, Lawrence Heights
•	Queen's Park, Ontario Provincial Government
•	Runnymede, Swansea
•	St. James Town
•	St. James Town, Cabbagetown
•	Steeles West, L'Amoreaux West
•	Stn A PO Boxes
•	Studio District
•	The Danforth West, Riverdale
•	Toronto Dominion Centre, Design Exchange
•	University of Toronto, Harbord
•	Westmount
•	Wexford, Maryvale
•	Willowdale, Willowdale East


### 5. Conclusion

The purpose of this project was to find a Neighborhood to open a Gym. We were finding a place with no other competitors. With the help of Data Science using data sources provided by Wikipedia and Foursquare, we could define each neighborhood and the venues it had. Then we analyzed the potential competition and end up with a list of possible neighborhoods.

The final decision will be made by the stakeholders based on the information provided and given the insight they have from other sources such as: average income, levels of attractiveness to sports activities, age, among others.
